cmd/autoretrieve/main.go

```diff
@@ -0,0 +1,775 @@
+
+				// ...start it on this goroutine
+				if err := r.retrieveFromBestCandidate(ctx, candidates); err != nil {
+					logger.Errorf("Could not retrieve %s: %v", entry.Cid, err)
+				} else {
+					logger.Infof("Successfully retrieved %v", entry.Cid)
+				}
+			}
+		}
+
+		if err := r.bsnet.SendMessage(ctx, sender, resMsg); err != nil {
+			logger.Errorf("Could not send bitswap message: %v", err)
+		}
+	}
+
+	// haveCount := len(resMsg.Haves())
+	// blockCount := len(resMsg.Blocks())
+	// dontHaveCount := len(resMsg.DontHaves())
+	// fmt.Printf("Finished bitswap message to %v (%v HAVE, %v BLOCK, %v DONTHAVE)\n", sender, haveCount, blockCount, dontHaveCount)
+}
+
+func (r *bsnetReceiver) ReceiveError(err error) {
+	//fmt.Printf("Bitswap receive error: %v\n", err)
+}
+
+func (r *bsnetReceiver) PeerConnected(id peer.ID) {
+	//fmt.Printf("Connected to peer %v\n", id)
+}
+
+func (r *bsnetReceiver) PeerDisconnected(id peer.ID) {
+	//fmt.Printf("Disconnected from peer %v\n", id)
+}
+
+type RetrievalCandidate struct {
+	Miner   address.Address
+	RootCid cid.Cid
+	DealID  uint
+}
+
+func GetRetrievalCandidates(endpoint string, c cid.Cid) ([]RetrievalCandidate, error) {
+
+	endpointURL, err := url.Parse(endpoint)
+	if err != nil {
+		return nil, xerrors.Errorf("endpoint %s is not a valid url", endpoint)
+	}
+	endpointURL.Path = path.Join(endpointURL.Path, c.String())
+
+	resp, err := http.Get(endpointURL.String())
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("http request to endpoint %s got status %v", endpointURL, resp.StatusCode)
+	}
+
+	var res []RetrievalCandidate
+
+	if err := json.NewDecoder(resp.Body).Decode(&res); err != nil {
+		return nil, xerrors.Errorf("could not unmarshal http response for cid %s", c)
+	}
+
+	return res, nil
+}
+
+type CandidateQuery struct {
+	Candidate RetrievalCandidate
+	Response  *retrievalmarket.QueryResponse
+}
+
+// Select the most preferable miner to retrieve from and execute the retrieval
+func (r *bsnetReceiver) retrieveFromBestCandidate(ctx context.Context, candidates []RetrievalCandidate) error {
+	checked := 0
+	var queries []CandidateQuery
+	var queriesLk sync.Mutex
+
+	var wg sync.WaitGroup
+	wg.Add(len(candidates))
+
+	for _, candidate := range candidates {
+		go func(candidate RetrievalCandidate) {
+			defer wg.Done()
+
+			query, err := r.fc.RetrievalQuery(ctx, candidate.Miner, candidate.RootCid)
+			if err != nil {
+				logger.Errorf(
+					"Failed to query retrieval %v/%v from miner %s for %s: %v",
+					checked+1,
+					len(candidates),
+					candidate.Miner,
+					candidate.RootCid,
+					err,
+				)
+				return
+			}
+
+			logger.Infof(
+				"Retrieval query %v/%v succeeded from miner %s for %s",
+				checked+1,
+				len(candidates),
+				candidate.Miner,
+				candidate.RootCid,
+			)
+
+			queriesLk.Lock()
+			queries = append(queries, CandidateQuery{Candidate: candidate, Response: query})
+			checked++
```

Comment:
instead of mutating / accessing the `checked` global inside of this goroutine, would it be safer to pass in the loop index as part of the goroutine call? i.e..
```go
for idx, candidate := range candidates {
  go func(idx uint, candidate RetrievalCandidate) {
    // .. retrieval logic
  }(idx, candidate)
}